abstract UA out, add futures support in

rjbs opened the pull request:
Okay, so probably this could be two commits, and maybe we can rewrite this commit later, but bear with me for now.

The end goal here is to allow the JMAP client to work in an async program, so that ->request can return a future that will resolve to the sentence collection when it's ready.  Meanwhile, the changes should not *seriously* break existing code.  So, it's okay if a small amount of mechanical change is needed for code that was calling things directly on the LWP object, but we shouldn't break every call to ->request in existing tests.

We accomplish this through a few steps.

First, we abstract out $tester->ua, previously a LWP::UserAgent object, into a thin adapter over the client of your choice.  LWP remains the default, but I've added a stub implementation for Net::Async::HTTP, which should work just fine, but will break logging.  (A few things in this commit worsen logging, some fatally for now.)

Next, we define the UA adapters to return futures, not responses.  The JMAP::Tester code that interprets those responses is then moved into a ->then sequence that hangs off the HTTP response.  Any user-facing method that works this way will either return a future or a Result, depending on the JMAP::Tester's return_futures setting.  The default is to return results, to match existing behavior.

If futures are returned, JMAP::Tester::Result::Failure objects show up as failures of the future.  If results are returned, Failure results may be returned, but other forms of failing futures will result in exceptions.

rjbs committed:
abstract out the UA role and allow futures

Okay, so probably this could be two commits, and maybe we can rewrite
this commit later, but bear with me for now.

The end goal here is to allow the JMAP client to work in an async
program, so that ->request can return a future that will resolve to the
sentence collection when it's ready.  Meanwhile, the changes should not
*seriously* break existing code.  So, it's okay if a small amount of
mechanical change is needed for code that was calling things directly on
the LWP object, but we shouldn't break every call to ->request in
existing tests.

We accomplish this through a few steps.

First, we abstract out $tester->ua, previously a LWP::UserAgent object,
into a thin adapter over the client of your choice.  LWP remains the
default, but I've added a stub implementation for Net::Async::HTTP,
which should work just fine, but will break logging.  (A few things in
this commit worsen logging, some fatally for now.)

Next, we define the UA adapters to return futures, not responses.  The
JMAP::Tester code that interprets those responses is then moved into
a ->then sequence that hangs off the HTTP response.  Any user-facing
method that works this way will either return a future or a Result,
depending on the JMAP::Tester's return_futures setting.  The default is
to return results, to match existing behavior.

If futures are returned, JMAP::Tester::Result::Failure objects show up
as failures of the future.  If results are returned, Failure results may
be returned, but other forms of failing futures will result in
exceptions.

wolfsage commented on lib/JMAP/Tester.pm:
```diff
@@ -78,6 +79,19 @@ There is also L<JMAP::Tester::Response/"as_stripped_pairs">.
+  default => 0,
+);
+
+my $Failsafe = sub {
```
Can we get some comments here to explain the purpose?

rjbs replied:
Yes!

When something doesn't work — not an individual method call, but the whole HTTP call, somehow — then the future will fail, and the failure might be a JMAP tester failure object, meaning we semi-expected it, or it might be some other crazy failure, meaning we had no way of seeing it coming.

We use Future->fail because that way we can use ->else in chains to only act on successful HTTP calls.  At the end, it's fine if you're expecting a future and can know that a failed future is a fail and a done future is okay.  In the old calling convention, though, you expect to get a success/fail object no matter what, and unexpected behavior is fatal.

`$Failsafe` emulates that.  Futures being resolved before return are converted like so:

* successful futures return their payload, the Result object
* failed futures that contain a JMAP tester Failure return the failure
* other failed futures die, like they would if you called `$failed_future->get`

wolfsage replied:
Thanks for the explanation! Can we get some condensed form of that as code comments in JMAP::Tester? (Perhaps my original comment wasn't entirely clear there!) :)

rjbs replied:
I've pushed up an explanation.

wolfsage commented on lib/JMAP/Tester.pm:
```diff
@@ -111,46 +125,13 @@ for my $type (qw(api authentication download upload)) {
   },
 );
 
-sub _set_cookie {
```
Removing this will break some downstream code I believe (like Ix, etc)

rjbs replied:
I wrote a branch in Ix, rjbs/future, that fixes this, already. :)

wolfsage commented on lib/JMAP/Tester.pm:
```diff
@@ -334,36 +315,25 @@ sub request {
-      http_request => $req,
-    });
-    return;
+  my $res_f = $self->ua->request($self, $post, jmap => {
```
I don't love the logging stuff, it's not clear the `jmap => {}` is for logging. Maybe just a matter of naming. `log_jmap` or something...

rjbs replied:
The logging stuff needs rethinking generally, for sure.  I think it's the most unclear WIP in this WIP.

wolfsage commented on lib/JMAP/Tester/UA/Async.pm:
```diff
@@ -0,0 +1,65 @@
+  my $logger = $tester->_logger;
+
+  my $log_method = "log_" . ($log_type // 'jmap') . '_request';
+  $self->ua->set_my_handler(request_send => sub {
```
Guessing `->ua` should be client since we have no `ua`?

rjbs replied:
Yes.  This whole class is… not clearly going to just work the first time. :-/

wolfsage commented on lib/JMAP/Tester/UA/Async.pm:
```diff
@@ -0,0 +1,65 @@
+    return;
+  });
+
+  my $http_res = $self->lwp->request($post);
```
I don't think we have an `lwp` either

rjbs replied:
This was very … um … a sketch of a library.  It's done now.

wolfsage commented on lib/JMAP/Tester/UA/Async.pm:
```diff
@@ -0,0 +1,65 @@
+
+  # Clear our handler, or it will get called for
+  # any http request our ua makes!
+  $self->ua->set_my_handler(request_send => undef);
```
more `ua` ..

wolfsage reviewed (commented):
I think the idea is okay. I'm not sure about your Net::Async::HTTP stub - I don't see that here and the Async stub that is included looks broken unless I'm misunderstanding something.

rjbs commented:
The Async code was a stub in the sense of "this is vaguely how it will look, but not done or tested."

If you think the idea is sound, I'll get at test run done. :)

rjbs committed:
Async: correctly delegate requests to async client

rjbs committed:
futures: explain what $Failsafe is for

rjbs commented:
I think that this is ready to go once we fix downstreams.

rjbs committed:
UA: require a set_default_header method

bad idea?

rjbs committed:
UA::LWP: do not call logging method that does not exist

this is meh

rjbs committed:
SentenceBroker: make immutable!

rjbs committed:
tests: remove weird and unneeded $broker

huh!

rjbs committed:
use a per-Response broker, with a backref

rjbs committed:
work to get auto-diagnostics on aborts

We need to twist the object graph around a bit so that when an abort is
being generated, we have access to the whole response.

Otherwise, an individual sentence is unable to throw an abort that shows
the enclosing response, which is probably what we want most of the time.

rjbs committed:
provide a pluggable diagnostics generator

* when no diagnostics are provided, provide some!
* let the provider be pluggable; starts with JSON, later Data::Printer?

The current pluggability of the diagnostician is not great.  I think it
should probably be something like:

1. you can replace the pretty printer from JSON to something else
2. you can, possibly per-abort, describe what diagnostics to generate

So maybe:

    $res = Result->new({
      ...,
      diagnostic_dumper => sub { ... },
      default_diagnostics => [
        'Response sentences' => sub ($res) { [ $res->items ] },
        ...
      ],
    });

Then diagnostics can be requested with:

    ->abort($message, [
      $label => $value,
      ...
    ]);

The second argument is a optlist (q.v. Data::OptList).  If there is no
value, the label is just a line.  If there is a value...

  ...if a subref, it's called with the $res as an argument.  The result
  is diagnostically dumpered.

  ...if another kind of ref, it's diagnostically dumpered.

We use OptList in part so that order is specified and not merely
cmp-based.

rjbs committed:
improved, more flexible diagnostics generator

rjbs committed:
allow subclassing to change default diag dumper

rjbs committed:
allow a Tester subclass to have its own Response class

wolfsage committed:
JMAP::Tester::UA::Async: Log requests just before sending them

This should very closely mimic what LWP::UserAgent's request_send hook
does which is log right before emitting the actual request.

There *might* be a chance that some other waiting object on the Loop
would run between when we log the request and when we emit it, but I'm
not sure, and if there was, I'd expect it to be a minimal delay...

wolfsage commented:
I think this is right now, based on downstream tests mostly passing :)

mmcclimon commented on lib/JMAP/Tester/SentenceBroker.pm:
```diff
@@ -40,4 +40,4 @@ sub strip_json_types {
 
 no Moo;
-1;
+__PACKAGE__->meta->make_immutable;
```
This doesn't actually _do_ anything, I think, because this is a Moo class, and Moo already generates immutable metaclasses.

rjbs replied:
You're probably right, but I think I'll leave it there unless you object.  Ending Moo-ish classes with make_immutable is probably better for my habits, and should be harmless.

mmcclimon commented on lib/JMAP/Tester.pm:
```diff
@@ -77,6 +79,19 @@ There is also L<JMAP::Tester::Response/"as_stripped_pairs">.
 
 =cut
 
+has return_futures => (
```
According to the ~Trurl-Klapaucius~ Rik Standard Naming Scheme, this should be `should_return_futures` or something. (`->return_futures` sounds a bit an accumulator of futures that you will, one day, return.)

rjbs replied:
Addressed in 0d9204e

mmcclimon commented on lib/JMAP/Tester.pm:
```diff
@@ -110,46 +125,13 @@ for my $type (qw(api authentication download upload)) {
-    }
-
-    return $ua;
+    require JMAP::Tester::UA::LWP;
+    JMAP::Tester::UA::LWP->new;
```
I'm putting this comment here because it's hard to comment on a commit generally:

I find this commit _extremely_ difficult to read. Your message says "Okay, so probably this could be two commits, and maybe we can rewrite this commit later, but bear with me for now," so I suppose I will bear with you for now, but I find it extremely difficult to follow what this commit is doing, which stinks because it's the most important commit in the entire branch.

rjbs replied:
You didn't provide the commit sha, so I am going to go dig it up.  Obviously, this should be GitHub's responsibility to do!  But things change, let us remember to include the commit sha when we need to point back to these.

rjbs replied:
This is 0d9204e.

rjbs replied:
I'm not sure how I would divide this commit usefully, but I'll give it some thought.

mmcclimon commented on lib/JMAP/Tester/UA/Async.pm:
```diff
@@ -43,22 +43,17 @@ sub request {
-  $self->ua->set_my_handler(request_send => undef);
-
-  return Future->done($http_res);
+  return $self->client->do_request(request => $req);
```
What even sort of thing _is_ a client? (I guess, probably, a Net::Async::HTTP object?) At the very least, I think you should stick in a comment that suggests what kind of thing you're expecting here.

rjbs replied:
It's a bad name.  I renamed it to `http_client`, which is what I have _usually_ used for Net::Async::HTTP objects, including in Synergy.

mmcclimon commented on lib/JMAP/Tester.pm:
```diff
@@ -84,6 +84,23 @@ has return_futures => (
   default => 0,
 );
 
+# When something doesn't work — not an individual method call, but the whole
+# HTTP call, somehow — then the future will fail, and the failure might be a
+# JMAP tester failure object, meaning we semi-expected it, or it might be some
+# other crazy failure, meaning we had no way of seeing it coming.
+#
+# We use Future->fail because that way we can use ->else in chains to only act
+# on successful HTTP calls. At the end, it's fine if you're expecting a future
+# and can know that a failed future is a fail and a done future is okay. In the
+# old calling convention, though, you expect to get a success/fail object no
+# matter what, and unexpected behavior is fatal.
+#
+# $Failsafe emulates that. Just before we return from a future-returning
+# method, and if the client is not set to return futures, we do this:
+#
+# * successful futures return their payload, the Result object
+# * failed futures that contain a JMAP tester Failure return the failure
+# * other failed futures die, like they would if you called $failed_future->get
```
You say that in the old calling convention, you expect to get a success/fail object no matter what (which is true). Even with $Failsafe, though, you can still get an exception, if there's some "other crazy failure"! This is counterintuitive, I think, based on the stated goals of $Failsafe, and I think it would make more sense to always generate a Failure object, even from some crazy failure.

I will admit that I don't exactly know what kind of failure condition would throw now when it didn't before, but: Failure objects are basically useless now (in that it's hard to get any useful information out of them, other than an http response), so making one for weird failures seems reasonable, unless I'm totally misunderstanding the intent of this code.

rjbs replied:
Right now, during `->request` in the main branch, we have this code (for example):
```
    $self->_logger->log_jmap_response({
      http_response => $http_res,
    });
```

If the _logger is missing somehow, or hasn't implemented `log_jmap_response`, or can't log the object in `http_res`, any of these could throw an exception, and `->request` would not catch it.  That behavior is preserved by $Failsafe.  The comment's "no matter what" is really weird:  it does X no matter what, except when it does Y.

Really, you get a Success for a 2xx response and a Failure for a non-2xx, and lots of other things throw.  I think that behavior is maintained.

mmcclimon commented on lib/JMAP/Tester/UA/LWP.pm:
```diff
@@ -59,17 +59,20 @@ sub set_default_header {
 sub request {
   my ($self, $tester, $req, $log_type, $log_extra) = @_;
 
+  Carp::cluck("something very strange happened") unless $tester->can('_logger');
```
You should make sure to `use Carp` in this file.

rjbs replied:
Thanks, done!

mmcclimon commented on lib/JMAP/Tester/UA/LWP.pm:
```diff
@@ -0,0 +1,78 @@
+}
+
+sub request {
+  my ($self, $tester, $req, $log_type, $log_extra) = @_;
```
I wonder: would it be better to stick, in Role::UA:

```perl
has tester => (
  is => 'ro',
  weak_ref => 1,
);
```

because saying `$self->ua->request($self, $request, ...)` feels very strange to me.

I guess you probably didn't do this so that you can pass some extra agent in from afar, so you can say 

```perl
my $http = My::Cool::HTTP->new;
my $ua = JMAP::Tester::UA::Async->new({ client => $client });
my $tester = JMAP::Tester->new({ ua => $ua });
```

So maybe it's ok, but I just wanted to mention that it feels weird to me.

mmcclimon replied:
Update: I have glanced at a couple of places that have to cope with the new code, and now I'm even more in favor of sticking a weak ref in, because `$tester->ua->request($tester, $req, 'foo')` feels dirty.

rjbs replied:
Can you elaborate on "couple of places"?  I think they were all "code using Tester has to dig into the UA to make a misc. HTTP call", which is now fixed by providing http_request and friends.  I don't mind setting up the weak ref, but I'd _rather_ not if we can agree that this makes it unneeded.

mmcclimon reviewed (commented):
I have made some comments. I guess it's fine, but it does feel kinda janky (as alh says, here or elsewhere) that everywhere we do stuff with the tester's LWP, we now break encapsulation.

That is: it's nice that we can now mechanically replace all instances of `$tester->ua` with `$tester->ua->lwp`, but that feels...something to me. Not "dangerous," exactly, but "grotty," maybe. It seems to me (I haven't yet reviewed all the branches for "cope with these changes") that the things we do with the tester's UA are fairly limited, and that it would be nicer to just abstract those two or three things, so that they mostly don't need to be changed. Of course, _some_  of them would, but mostly they wouldn't. (If you replaced the ua role's `set_default_header` with `default_header`, for example, this would eliminate a bunch of required mechanical changes.)

rjbs committed:
LWP agent: ensure that set_default_header returns ()

rjbs committed:
add scan_cookies to UA interface

rjbs committed:
Async UA: rename ->client to ->http_client

rjbs committed:
JMAP::Tester: add ->http_* methods

These are here to provide a uniform interface to making HTTP
requests using the UA, whatever it might be, even when not
making a JMAP method-call request.

rjbs committed:
rename return_futures to should_return_futures

...and add docs

rjbs committed:
LWP UA: since we use Carp, require it!

rjbs committed:
UA: add get_default_header

In LWP this is part of ->header, but it is sort of byzantine.

rjbs committed:
LWP UA: set C-T to application/json by default

rjbs committed:
drop make_immutable (it does nothing in Moo)

and was only in one place, anyway

rjbs committed:
Logger: handle "misc" request/response

mmcclimon commented:
88173d1 has an unrelated docs tweak for $Failsafe, which could be squashed into baac7a2e.

wolfsage commented on lib/JMAP/Tester.pm:
```diff
@@ -839,26 +875,99 @@ sub logout {
+
+This method respects the C<should_return_futures> attributes of the
+JMAP::Tester object, and in futures mode will return a future that will resolve
+to the HTTP::Response.
```
We document HTTP::Response here which is nice but we don't spell out that $http_requeset is an HTTP::Request object, and I think we should.

It'd be nice if we also L<...> wrapped them

rjbs replied:
Updated.

wolfsage commented on lib/JMAP/Tester/UA/LWP.pm:
```diff
@@ -19,6 +19,8 @@ has lwp => (
     $lwp->cookie_jar({});
 
+    $lwp->default_header('Content-Type' => 'application/json');
+
```
Shouldn't this also be charset utf8

Also, why are we only doing this on the LWP UA and not also the Async UA?

wolfsage committed:
Async UA: Net::Async::HTTP::Client doesn't have coookie_jar meth

So we need to reach into the object to use it

wolfsage committed:
Async UA: Set Content-Type to application/json by default

wolfsage committed:
Async UA: Don't try to use log methods that don't exist

... on say, the Null logger

rjbs committed:
docs: explain the argument to http_request
